File: rust/2024/src/days/day02.rs
use crate::advent_utils::Solution as AdventSolution;

pub struct Solution;

type Validator = fn(&Vec<i32>, fn(&Vec<i32>) -> Vec<i32>) -> bool;

impl AdventSolution for Solution {
    /// ```
    /// # use aoc2024::days;
    /// # use aoc2024::Solution as _;
    /// let input = std::fs::read_to_string("../../tests/2024/day02.txt").unwrap();
    /// assert_eq!(days::day02::Solution::part_1(&input), 2);
    /// ```
    fn part_1(data: &str) -> usize {
        data.lines()
            .map(|l| {
                l.split_whitespace()
                    .filter_map(|s| s.parse::<i32>().ok())
                    .collect::<Vec<_>>()
                    .windows(2)
                    .map(|w| w[0] - w[1])
                    .collect::<Vec<_>>()
            })
            .filter(|l| {
                !l.is_empty()
                    && (l.iter().all(|v| v.abs() <= 3))
                    && (l.iter().all(|&i| i < 0) || l.iter().all(|&i| i > 0))
            })
            .count()
    }
    /// ```
    /// # use aoc2024::days;
    /// # use aoc2024::Solution as _;
    /// let input = std::fs::read_to_string("../../tests/2024/day02.txt").unwrap();
    /// assert_eq!(days::day02::Solution::part_2(&input), 4);
    /// let input2 = std::fs::read_to_string("../../tests/2024/day022.txt").unwrap();
    /// assert_eq!(days::day02::Solution::part_2(&input2), 28);
    /// ```
    fn part_2(data: &str) -> usize {
        data.lines()
            .filter(|l| {
                let diff: fn(&Vec<i32>) -> Vec<i32> =
                    |lst| lst.windows(2).map(|sl| sl[0] - sl[1]).collect();
                let validate: Validator = |lst, diff| {
                    let df = diff(lst);
                    df.iter().all(|&v| (1..=3).contains(&v.abs()))
                        && (df.iter().all(|v| v.is_positive())
                            || df.iter().all(|v| v.is_negative()))
                };

                let mut nums: Vec<i32> = l
                    .split_ascii_whitespace()
                    .filter_map(|s| s.parse().ok())
                    .collect();

                if validate(&nums, diff) {
                    return true;
                }

                let difs = diff(&nums);
                let mags = difs
                    .iter()
                    .map(|&v| ((1..=3).contains(&v.abs()), v > 0, v < 0))
                    .collect::<Vec<_>>();
                let pos = mags.iter().filter(|(_, b, _)| *b).count();
                let neg = mags.iter().filter(|(_, _, c)| *c).count();
                let failed = mags
                    .into_iter()
                    .enumerate()
                    .find(|&(_, (a, b, c))| !(a && (if pos > neg { b } else { c })))
                    .unwrap()
                    .0;
                let mut v1 = nums.clone();
                v1.remove(failed);
                nums.remove((failed + 1).min(nums.len() - 1));
                if validate(&v1, diff) || validate(&nums, diff) {
                    return true;
                };

                false
            })
            .count()
    }
}

File: python/src/2024/days/day02.py
"""day02."""

MIN_DIF, MAX_DIF = 1, 3


class Solution:
    """Solution for day 02."""

    @staticmethod
    def part_1(data: str) -> int:
        """
        Part 1 solution.

        ```python3
        >>> f = open("../tests/2024/day02.txt")
        >>> Solution.part_1(f.read())
        2

        ```

        Returns
        -------
        The result.

        """
        lnums = (line.split() for line in data.splitlines())
        difs = [
            [int(nums[i]) - int(nums[i + 1]) for i in range(len(nums) - 1)]
            for nums in lnums
        ]
        return sum(
            all(MIN_DIF <= abs(d) <= MAX_DIF for d in difc)
            and (all(d < 0 for d in difc) or all(d > 0 for d in difc))
            for difc in difs
        )

    @staticmethod
    def part_2(data: str) -> int:
        """
        Part 2 solution.

        ```python3
        >>> f = open("../tests/2024/day02.txt")
        >>> Solution.part_2(f.read())
        4
        >>> f2 = open("../tests/2024/day022.txt")
        >>> Solution.part_2(f2.read())
        28
        >>> f.close(); f2.close()

        ```

        Returns
        -------
        The result.

        """
        lnums = [[int(n) for n in line.split()] for line in data.splitlines()]
        invalids = [(lst, _calc_difs(lst)) for lst in lnums if not _validate(lst)]
        checks = [
            (
                nums,
                [MIN_DIF <= abs(d) <= MAX_DIF for d in difs],
                [i > 0 for i in difs],
                [i < 0 for i in difs],
            )
            for nums, difs in invalids
        ]
        checks_merged = [
            (nums, mag, pos if sum(pos) > sum(neg) else neg)
            for nums, mag, pos, neg in checks
        ]
        wrong_indices = [
            (nums, next(i for i in range(len(nums) - 1) if not (mag[i] and sign[i])))
            for nums, mag, sign in checks_merged
        ]

        variations = [
            (lst, _remove_item(lst, idx), _remove_item(lst, idx + 1))
            for lst, idx in wrong_indices
        ]

        validated = [(lst, _validate(a), _validate(b)) for lst, a, b in variations]
        rem = [not frst and not scnd for _, frst, scnd in validated]

        return len(lnums) - sum(rem)


def _calc_difs(nums: list[int]) -> list[int]:
    return [nums[i] - nums[i + 1] for i in range(len(nums) - 1)]


def _validate(nums: list[int]) -> bool:
    difc = _calc_difs(nums)
    return all(MIN_DIF <= abs(d) <= MAX_DIF for d in difc) and (
        all(d < 0 for d in difc) or all(d > 0 for d in difc)
    )


def _remove_item(nums: list[int], idx: int) -> list[int]:
    new = nums.copy()
    idx_checked = min(len(nums) - 1, idx)
    _ = new.pop(idx_checked)
    return new
